### PR TITLE
Stabilize asm! and global_asm!

### PR DESCRIPTION
**Pull request by Amanieu**
Tracking issue: #72016

It's been almost 2 years since the original [RFC](https://github.com/rust-lang/rfcs/pull/2850) was posted and we're finally ready to stabilize this feature!

The main changes in this PR are:
- Removing `asm!` and `global_asm!` from the prelude as per the decision in #87228.
- Stabilizing the `asm` and `global_asm` features.
- Removing the unstable book pages for `asm` and `global_asm`. The contents are moved to the [reference](https://github.com/rust-lang/reference/pull/1105) and [rust by example](https://github.com/rust-lang/rust-by-example/pull/1483).
  - All links to these pages have been removed to satisfy the link checker. In a later PR these will be replaced with links to the reference or rust by example.
- Removing the automatic suggestion for using `llvm_asm!` instead of `asm!` if you're still using the old syntax, since it doesn't work anymore with `asm!` no longer being in the prelude. This only affects code that predates the old LLVM-style `asm!` being renamed to `llvm_asm!`.
- Updating `stdarch` and `compiler-builtins`.
- Updating all the tests.

r? @joshtriplett

### CODE REVIEW

File: Cargo.lock [hunk 1]
```diff
@@ -680,9 +680,9 @@ dependencies = [
 
 [[package]]
 name = "compiler_builtins"
-version = "0.1.55"
+version = "0.1.65"
```
Comment (Amanieu) [inline Cargo.lock:683]: In case anyone's wondering: yes, I made a typo when publishing a new version of compiler-builtins.

**Comment by nbdd0121** [comment]
FCP is already completed in the tracking issue.

**Comment by camelid** [comment]
Ah, I thought I checked, not sure how I missed that.

File: compiler/rustc_lint/src/context.rs [hunk 2]
```diff
@@ -791,7 +791,7 @@ pub trait LintContext: Sized {
                 }
                 BuiltinLintDiagnostics::NamedAsmLabel(help) => {
                     db.help(&help);
-                    db.note("see the asm section of the unstable book <https://doc.rust-lang.org/nightly/unstable-book/library-features/asm.html#labels> for more information");
+                    //db.note("see the asm section of the unstable book <https://doc.rust-lang.org/nightly/unstable-book/library-features/asm.html#labels> for more information");
```
Comment (joshtriplett) [inline compiler/rustc_lint/src/context.rs:794]:
```suggestion
```

File: library/std/src/sys/solid/abi/mod.rs [hunk 3]
```diff
@@ -10,9 +10,9 @@ pub fn breakpoint_program_exited(tid: usize) {
         match () {
             // SOLID_BP_PROGRAM_EXITED = 15
             #[cfg(target_arch = "arm")]
-            () => asm!("bkpt #15", in("r0") tid),
-            #[cfg(target_arch = "aarch64")]
-            () => asm!("hlt #15", in("x0") tid),
+            () => core::arch::asm!("bkpt #15", in("r0") tid),
+            #[cfg(core::arch::asm = "aarch64")]
```
Comment (joshtriplett) [inline library/std/src/sys/solid/abi/mod.rs:14]:
```suggestion
            #[cfg(target_arch = "aarch64")]
```

File: src/doc/unstable-book/src/compiler-flags/sanitizer.md [hunk 4]
```diff
@@ -199,7 +199,7 @@ LLVM CFI can be enabled with -Zsanitizer=cfi and requires LTO (i.e., -Clto).
 ## Example
 
 ```text
-#![feature(asm, naked_functions)]
+#![feature(naked_functions)]
 
 use std::mem;
```
Comment (joshtriplett) [inline src/doc/unstable-book/src/compiler-flags/sanitizer.md:204]:
```suggestion
use std::arch::asm;
use std::mem;
```

File: src/test/assembly/x86_64-fortanix-unknown-sgx-lvi-inline-assembly.rs [hunk 5]
```diff
@@ -26,11 +26,13 @@ pub extern fn get(ptr: *const u64) -> u64 {
...
-pub extern fn myret() {
+pub extern "C" fn myret() {
     unsafe {
-        asm!(".start_myret_inline_asm:
+        asm!(
+            ".start_myret_inline_asm:
             ret
-            .end_myret_inline_asm:");
+            .end_myret_inline_asm:"
+        );
```
Comment (joshtriplett) [inline src/test/assembly/x86_64-fortanix-unknown-sgx-lvi-inline-assembly.rs:35]: If you're going to change the formatting of this anyway, consider changing it to:
```rust
        asm!(
            ".start_myret_inline_asm:",
            "ret",
            ".end_myret_inline_asm:",
        );
```

File: src/test/ui/macros/macros-nonfatal-errors.stderr [hunk 6]
```diff
@@ -221,5 +215,14 @@ error: trace_macros! accepts only `true` or `false`
 LL |     trace_macros!(invalid);
    |     ^^^^^^^^^^^^^^^^^^^^^^
 
+error: cannot find macro `asm` in this scope
```
Comment (joshtriplett) [inline src/test/ui/macros/macros-nonfatal-errors.stderr:218]: This test needs to `use std::arch::asm;` so that it tests what it's intending to test.

**Comment by joshtriplett** [comment]
I just posted a set of review comments; with all of those addressed, r=me.

File: compiler/rustc_lint_defs/src/builtin.rs [hunk 7]
```diff
@@ -2455,7 +2455,7 @@ declare_lint! {
     ///
     /// ### Example
     ///
-    /// ```rust
+    /// ```rust,ignore (fails on non-x86_64)
     /// #[cfg(target_arch="x86_64")]
```
Comment (falaosu) [inline compiler/rustc_lint_defs/src/builtin.rs:2459]:
```suggestion
    /// #![cfg(target_arch="x86_64")]
```

Comment (Amanieu) [reply compiler/rustc_lint_defs/src/builtin.rs:2459]: That won't work. The tests are checking that the lint actually triggers in the example, but it only triggers on x86_64.

**Comment by Andy-Python-Programmer** [comment]
Yay! Now we are one step closer of compiling operating systems in stable rust :^)

**Comment by rylev** [comment]
This seems likely to be noise as a result of the issue tracked in https://github.com/rust-lang/rustc-perf/issues/1126 -- marking as triaged, not something we should address directly.

@rustbot label +perf-regression-triaged

**Comment by FCLC** [comment]
[x86_64, AVX512]
I might be missing something in the docs, but AFAICT as of now the only supported avx512 instruction sets are the foundation set (avx512F) and the Byte and Word set (avx512BW), the implication being that the V(ector) L(ength) and D(ual) Q(uad) sets are not currently supported. 

Reason I bring this up is mainly to do with the exclusion of AVX512VL. The VL set extends the foundation set (and most other sets) to be able to operate on both XMM and YMM registers, allowing for much cleaner implementations as well as faster overall code. 

The reason for wanting this functionality is due the lower latency of certain AVX and AVX2 instructions that can only operate on xmm registers, but wanting the more modern and advanced instructions available under AVX512. 

For example, as of now if we wanted to do a fuse multiply add on a group of 4 packed singles without avx512VL we'd have to do something along the lines of 

`
xor/zero out all your registers 
`
`vmov zmm1, {mem pointer to 4 PS} `
`vfmadd132 zmm1, zmm1, zmm1; do a fuse multiply add on each of the 4 packed singles on themselves`
`vextractf32x4 xmm1, zmm1, 0`
`whatever you need to do in xmm registers`
`vinsertf32x4 zmm1, zmm1, xmm1, 0 ; we're back in ZMM registers for any avx512 instruction work we need to do` 


The VL set of extensions allows us to use AVX512 instructions directly on xmm and ymm registers, removing the need for copying back and forth between AVX1/2 registers and AVX512 registers. 

This is also relevant not only because of the latency cost, but also because they may find themselves on different ports of the CPU core itself, slowing us down further. 

The above would become:

`vmov xmm1, {mem pointer to 4 PS} `
`vfmadd132 xmm1, xmm1, xmm1; do a fuse multiply add on each of the 4 packed singles on themselves`
`{other SIMD instructions from AVX1/2 512 or SSE all operating on xmm without extra copies}`

**Pull request merged**